src/persist/src/sqlite.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Implementation of [Consensus] backed by sqlite.
+
+use std::path::Path;
+use std::sync::Arc;
+use std::time::Instant;
+
+use async_trait::async_trait;
+use rusqlite::{named_params, params, Connection, OptionalExtension};
+use tokio::sync::Mutex;
+
+use crate::error::Error;
+use crate::storage::{Consensus, SeqNo, VersionedData};
+
+const APPLICATION_ID: i32 = 0x0678_ef32; // chosen randomly
+
+const SCHEMA: &str = "
+CREATE TABLE consensus (
+    shard text NOT NULL UNIQUE,
+    sequence_number integer NOT NULL,
+    data blob NOT NULL
+);
+";
+
+/// Implementation of [Consensus] over a sqlite database.
+#[derive(Debug)]
+pub struct SqliteConsensus {
+    conn: Arc<Mutex<Connection>>,
+    shard: String,
+}
+
+impl SqliteConsensus {
+    /// Open a sqlite-backed [Consensus] instance at `path`, for the collection
+    /// named `shard`.
+    pub fn open(path: &Path, shard: String) -> Result<Self, Error> {
```

Comment:
nit: do the `AsRef<Path>` trick here, it seems to be idiomatic for this sort of thing

Reply:
oh another thought that perhaps makes this last comment irrelevant. my working assumption is that in a single-binary deployment (for those following along, the only place where this is used), then this would share a sqlite instance with the catalog. in which case, we may want `open` to accept a connection instead of a path (the catalog requires that the sqlite instance is opened exclusively)

Reply:
hmmm my working assumption was that this would be a separate sqlite instance from the catalog, the way that stash instances are separate sqlite instances from the catalog. I wanted to originally build this on top of stash but it seemed stash really wanted to present as a differential collection (which could be made to work i guess but seemed complex).

Reply:
Discussed with Ruchir in the office, but sharing a SQLite connection among threads leads to `SQLITE_BUSY` exceptions and such. It's generally proven much nicer to let each use manage its own SQLite database(s). `Consensus` feels fundamentally different enough from `Stash` that seems ok and maybe even good to have them separate.

Reply:
i'm convinced! separate sgtm

Reply:
did the AsRef<Path> thing!